### ISSUE TITLE
Shapely 1.7.0 error on intersect with invalid geometries

### ISSUE THREAD

**Issue by ger-benjamin** [openoereb/pyramid_oereb#1013]
Hi,

It concerns pyramid_oereb master or dev build after: https://github.com/openoereb/pyramid_oereb/pull/1003

We have moved to `Shapely 1.7.0`. I've tried to update a pyramid_oereb with it but it looks crashing on big parcels with this error:

```
TopologicalError: The operation 'GEOSIntersection_r' could not be performed. Likely cause is invalidity of the geometry <shapely.geometry.polygon.Polygon object at 0x7f8007b1f050>
```

Here: `File .../oereb/.venv/lib/python2.7/site-packages/pyramid_oereb/lib/records/geometry.py", line 134.`

Perhaps I get this error because I still use python 2.7 in the project ? **Can someone does some tests on big parcels to see if the error happens too ?**
I wonder if it is always happening or only without python 3

If nobody can reproduce the issue, I'll try to export the polygon to test on my side and maybe open an issue in the Shapely repo.

```
.../oereb/.venv/lib/python2.7/site-packages/pyramid_oereb/lib/records/view_service.py:119: UserWarning: Type of "layer_opacity" should be "float"
  warnings.warn('Type of "layer_opacity" should be "float"')
ERROR shapely.geos TopologyException: Input geom 0 is invalid: Ring Self-intersection at or near point 2699116.8500000001 1208887.851 at 2699116.8500000001 1208887.851
mod_wsgi (pid=9903): Exception occurred processing WSGI script '.../oereb/apache/application.wsgi'.
Traceback (most recent call last):
  File .../oereb/.venv/lib/python2.7/site-packages/pyramid/router.py", line 270, in __call__
    response = self.execution_policy(environ, self)
  File .../oereb/.venv/lib/python2.7/site-packages/pyramid/router.py", line 279, in default_execution_policy
    return request.invoke_exception_view(reraise=True)
  File .../oereb/.venv/lib/python2.7/site-packages/pyramid/view.py", line 778, in invoke_exception_view
    reraise_(*exc_info)
  File .../oereb/.venv/lib/python2.7/site-packages/pyramid/router.py", line 277, in default_execution_policy
    return router.invoke_request(request)
  File .../oereb/.venv/lib/python2.7/site-packages/pyramid/router.py", line 249, in invoke_request
    response = handle_request(request)
  File .../oereb/.venv/lib/python2.7/site-packages/pyramid/tweens.py", line 43, in excview_tween
    response = _error_handler(request, exc)
  File .../oereb/.venv/lib/python2.7/site-packages/pyramid/tweens.py", line 17, in _error_handler
    reraise(*exc_info)
  File .../oereb/.venv/lib/python2.7/site-packages/pyramid/tweens.py", line 41, in excview_tween
    response = handler(request)
  File .../oereb/.venv/lib/python2.7/site-packages/pyramid/router.py", line 148, in handle_request
    registry, request, context, context_iface, view_name
  File .../oereb/.venv/lib/python2.7/site-packages/pyramid/view.py", line 667, in _call_view
    response = view_callable(context, request)
  File .../oereb/.venv/lib/python2.7/site-packages/pyramid/config/views.py", line 188, in attr_view
    return view(context, request)
  File .../oereb/.venv/lib/python2.7/site-packages/pyramid/config/views.py", line 214, in predicate_wrapper
    return view(context, request)
  File .../oereb/.venv/lib/python2.7/site-packages/pyramid_oereb/contrib/stats/decorators.py", line 9, in wrapper
    response = wrapped(context, request)
  File .../oereb/.venv/lib/python2.7/site-packages/pyramid/viewderivers.py", line 401, in viewresult_to_response
    result = view(context, request)
  File .../oereb/.venv/lib/python2.7/site-packages/pyramid/viewderivers.py", line 116, in _class_requestonly_view
    response = getattr(inst, attr)()
  File .../oereb/.venv/lib/python2.7/site-packages/pyramid_oereb/views/webservice.py", line 262, in get_extract_by_id
    self._request.route_url('{0}/sld'.format(route_prefix))
  File .../oereb/.venv/lib/python2.7/site-packages/pyramid_oereb/lib/processor.py", line 287, in process
    extract = self.plr_tolerance_check(extract_raw)
  File .../oereb/.venv/lib/python2.7/site-packages/pyramid_oereb/lib/processor.py", line 95, in plr_tolerance_check
    if public_law_restriction.calculate(real_estate):
  File .../oereb/.venv/lib/python2.7/site-packages/pyramid_oereb/lib/records/plr.py", line 186, in calculate
    self.length_unit, self.area_unit
  File .../oereb/.venv/lib/python2.7/site-packages/pyramid_oereb/lib/records/geometry.py", line 134, in calculate
    result = self._extract_collection(self.geom.intersection(real_estate.limit))
  File .../oereb/.venv/lib/python2.7/site-packages/shapely/geometry/base.py", line 649, in intersection
    return geom_factory(self.impl['intersection'](self, other))
  File .../oereb/.venv/lib/python2.7/site-packages/shapely/topology.py", line 70, in __call__
    self._check_topology(err, this, other)
  File .../oereb/.venv/lib/python2.7/site-packages/shapely/topology.py", line 38, in _check_topology
    self.fn.__name__, repr(geom)))
TopologicalError: The operation 'GEOSIntersection_r' could not be performed. Likely cause is invalidity of the geometry <shapely.geometry.polygon.Polygon object at 0x7f8007b1f050>
```

**Comment by jwkaltz** [comment]
It seems several cantons have now used the new pyramid_oereb version, with the new shapely version, and have not encountered this problem. But these, as far as I know, are using python3.

@kdeininger Are you able to test as well?

**Comment by kdeininger** [comment]
I have encountered the same issue in another project after upgrading Shapely. That project uses Python 3, so I do not think, that it is related to the Python version.

As the error log shows, the geometry seems to be invalid.

```
ERROR shapely.geos TopologyException: Input geom 0 is invalid: Ring Self-intersection at or near point 2699116.8500000001 1208887.851 at 2699116.8500000001 1208887.851
```

This is also the case in my other project. I could not find any hint in the changelog, but somehow the new Shapely version is less tolerant.

The best solution would be to repair the geometry, but as a quick fix, downgrading Shapely (<1.7) worked in my other project.

**Comment by CoNussberger** [comment]
I encountered the error when testing pyramid_oereb Verstion 1.7.3 with python 3.6 and shaply 1.7.
I attached a extract of our oereb_error log, where you can see how the error is manifesting.
[oereb_error.log](https://github.com/openoereb/pyramid_oereb/files/4420130/oereb_error.log)

Question: Is it possible to use pyramit_oereb Version 1.7.3. with a lower version of shaply?

For the oereb_server requirements file i use the following content:
c2c.template==2.1.0
pyramid_oereb[recommend]==1.7.3
-e .

I guess i have to change this to uese al lower version of shaply if possible.

**Comment by jwkaltz** [comment]
@CoNussberger I recommend you try to fix the data, it appears that the new version of shapely is more strict regarding data correctness:
TopologyException: Input geom 0 is invalid: Ring Self-intersection at or near point 2773227.6068999991 1201353.8407999985 at 2773227.6068999991 1201353.8407999985

At Schwyz, we will also try to fix the data, as a temporary workaround, we downgraded the shapely version, but to do that, you need to enter all requirements of pyramid_oereb into your project-specific requirements instead, which is not sustainable. So: please give priority to data fixing :)

**Comment by CoNussberger** [comment]
@jwkaltz Thanks  for your recommendation. I will try to fix the data, but as it is concerning communal data, that is not going to be as straight forward. So I also use the temporary workaround of downgrading Shaply.
To downgrade Shapely I found it is sufficient to just add the Version of Shapely to the requirements file like that:

c2c.template==2.1.0
Shapely==1.6.4
pyramid_oereb[recommend]==1.7.3
-e .

**Comment by jwkaltz** [comment]
@CoNussberger @kdeininger Regarding this issue, using the example from Schwyz data, we have found out that the cause was that the geometry data of the public law restriction was invalid. It seems that shapely 1.6 does not report an error when it is doing an intersect of a parcel geometry with a not valid public law restriction geometry, but shapely 1.7 is more strict here and does report an error.

Can you please check in your cases the validity of the data (the public law restriction geometry data)?
To me, it does not seem safe to continue using shapely 1.6, I am not sure whether the concerned plrs for a parcel are being determined correctly if the plr geometry is not correct.

**Comment by RenataMuellerC2C** [comment]
To reproduce the error in pyramid-oereb:

1. Replace the geometry1 (multipolygon) in file real_estates.json as ‘limit’
2. Replace geometry2 (polygon, public_law_restriction) in file geomtry.json as 'limit'
3. Set the shapely version in requirements.txt  to 1.7.0 (if its not)
4. In browser call http://localhost:6543/oereb/extract/reduced/xml/CH113928077734

[geomety1.txt](https://github.com/openoereb/pyramid_oereb/files/4580793/geomety1.txt)
[geometry2.txt](https://github.com/openoereb/pyramid_oereb/files/4580794/geometry2.txt)

**Comment by jwkaltz** [comment]
For information, Canton Schwyz is discussing with Swisstopo the issue of inconsistency between Interlis Geometry validity vs. OGC geometry validity (this is in reality the source of the problem described in this ticket), to find out how it should be handled in general.

**Comment by jwkaltz** [comment]
I am closing this issue, as the fundamental question should now be addressed in a dedicated issue, #1111
